File: java/876_Middle_of_the_Linked_List/876_Middle_of_the_Linked_List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode middleNode(ListNode head) {
        int length = nodeCount(head);
        int middle = length/2;
        boolean selectSecond = length % 2 == 0 ? false : true;
        
        int pos = 1;
        ListNode curr = head;
        while (pos < middle) {
            curr = curr.next;
            pos++;
        }
        return selectSecond ? curr.next : curr;
    }
    
    private int nodeCount(ListNode head) {
        ListNode curr = head;
        int total = curr == null ? 0 : 1;
        while(curr != null) {
            curr = curr.next;
            total++;
        }
        return total;
    }
}

File: go/876_Middle_of_the_Linked_List/876_Middle_of_the_Linked_List.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func middleNode(head *ListNode) *ListNode {
    length := nodeCount(head)
    middle := length/2
    selectSecond := false
    
    if length & 1 == 1 {
        selectSecond = true
    }
    pos := 1
    cur := head
    for pos < middle {
        cur = cur.Next
        pos++
    }
    if selectSecond {
        return cur.Next
    } else {
        return cur
    }
}

func nodeCount(head *ListNode) int {
    cur := head
    total := 1
    if cur == nil {
        total = 0
    }
    for cur != nil {
        cur = cur.Next
        total++
    }
    return total
}
